### PR TITLE
Implementation of a trie data structure for residents, towns, and nations

### PR DESCRIPTION
**Pull request by stzups**
#### Description: 

Thanks to @suneettipirneni for suggesting using a trie

This pull request adds a trie structure which is built asynchronously upon the loading of Towny. It makes searching through many residents, towns, and nations for tab complete suggestions nearly instant. One test was conducted with 20,000 nations, and fetching completions with one character, for example typing "a", took 7ms. Fetching with two characters or more is nearly instant (>1ms).

 
Changes include:
- Added Trie.java
- Added residentsTrie, townsTrie, and nationsTrie to TownyUniverse 
- All tries are built asynchronously when TownyUniverse is loaded
- New methods in BaseCommand.java: 
getTownyStartingWith() and filterByStartOrGetTownyStartingWith() 
- Removal of unnecessary methods in NameUtil.java
- Removal of MAX_RETURNS because the only usage is with possible subcommands (e.g. nationTabCompletes), not residents, towns, or nations which could return hundreds of results

____
- [x] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

**Comment by LlmDl** [comment]
Are you on our [discord](https://discord.gg/gnpVs5m) or [irc](http://webchat.esper.net/?channels=towny) yet?

**Comment by suneettipirneni** [comment]
This looks really good, the only thing I think thats should be added is repeating task that regenerates the tries every x minutes. Otherwise nice job!

**Comment by stzups** [comment]
Removal/renames of residents, towns, and nations have been added. Trie has been optimized and also only returns the first 100 results, making tab completions nearly instant. Here is a summary of the tests: 

Each line is the tab completion of a string, with the first line being blank ("") and the third line being 2 characters ("ab"), and so on. The first line ("") searches through all 200,000 nations unless a limit was imposed. 

Old search which went through all entries, 200k nations
image

Old search with limit of 20 results, 200k nations
image
The search times go up because more nations have to be searched through to find at least 20. If there are less than 20 nations, the entire nation map had to be searched.

Trie, no limit, 200k nations
image

Trie, limit of 100, 200k nations (final version)
image

Building a trie with 200k nations takes roughly 1.5-2ms, which will only occur asynchronously on load

File: src/com/palmergames/bukkit/towny/command/BaseCommand.java [hunk 1]
```diff
@@ -1,55 +1,75 @@
...
+			matches.addAll(NameUtil.filterByStart(NameUtil.getNames(townyUniverse.getWorldMap().values()), arg));
+		}
 
-			}
-			// Match towns
-			for (Town town : townyUniverse.getDataSource().getTowns()) {
-				if (town.getName().toLowerCase().startsWith(lastArg)) {
-					output.add(town.getName());
-				}
+		//System.out.println("Found "+matches.size()+" for "+type+" in "+(float)(System.nanoTime()-start)/1000000+"ms");
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/command/BaseCommand.java]: Remove debug code.

File: src/com/palmergames/util/Trie.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		// Current trieNode to crawl through
+		TrieNode trieNode = root;
+		Queue<TrieNode> found = Collections.asLifoQueue(new LinkedList<>());
+
+		// Loop through each character of key
+		for (int i = 0; i < key.length(); i++) {
+			char index = Character.toLowerCase(key.charAt(i)); // Case insensitive
+
+			TrieNode lastNode = trieNode;
+			//found.add(lastNode); // Adds to list of TrieNodes to remove
```
Comment (LlmDl) [inline src/com/palmergames/util/Trie.java]: Remove this line?

**Pull request merged**